### PR TITLE
[SPARK-3586][streaming]Support nested directories in Spark Streaming

### PR DESCRIPTION
**Pull request by wangxiaojing**
For text files, the method streamingContext.textFileStream(dataDirectory). 
The improvement of the streaming to Support subdirectories,spark streaming can  monitor the subdirectories dataDirectory and process any files created in that directory.
eg:
streamingContext.textFileStream(/test). 
Look at the direction contents:
/test/file1
/test/file2
/test/dr/file1
if the directory "/test/dr/" have new file "file2" ,spark streaming can process  the file

### CODE REVIEW

**Comment by tdas** [comment]
Jenkins this is OK to test

**Comment by tdas** [comment]
@wangxiaojing This PR has merge conflicts. Could you fix them?

**Comment by wangxiaojing** [comment]
@tdas Thanks a lot . I fix them.

**Comment by wangxiaojing** [comment]
@tdas rebase the latest master and update.

File: streaming/src/main/scala/org/apache/spark/streaming/StreamingContext.scala [hunk 1]
```diff
@@ -327,6 +327,7 @@ class StreamingContext private[streaming] (
    * Files must be written to the monitored directory by "moving" them from another
    * location within the same file system. File names starting with . are ignored.
```
Comment (tdas) [inline streaming/src/main/scala/org/apache/spark/streaming/StreamingContext.scala]: Please update JavaStreamingContext similarly

**Comment by wangxiaojing** [comment]
@tdas

**Comment by wangxiaojing** [comment]
@tdas

**Comment by srowen** [comment]
@wangxiaojing I'd like to revive this PR and get it committed. There have been a number of requests for this functionality, and several JIRAs and PRs about it. Would you be willing to start by rebasing? It might even be easier if you squash first on your end. I will leave some more comments.

File: docs/streaming-programming-guide.md [hunk 2]
```diff
@@ -641,17 +641,17 @@ methods for creating DStreams from files and Akka actors as input sources.
 
     <div class="codetabs">
     <div data-lang="scala" markdown="1">
-        streamingContext.fileStream[KeyClass, ValueClass, InputFormatClass](dataDirectory)
+        streamingContext.fileStream[KeyClass, ValueClass, InputFormatClass](dataDirectory, depth)
```
Comment (srowen) [inline docs/streaming-programming-guide.md]: I think all of the examples should continue to show this without the `depth` parameter, since it's optional and for particular use cases.

File: docs/streaming-programming-guide.md [hunk 3]
```diff
@@ -641,17 +641,17 @@ methods for creating DStreams from files and Akka actors as input sources.
...
+		streamingContext.fileStream<KeyClass, ValueClass, InputFormatClass>(dataDirectory, depth);
     </div>
     <div data-lang="python" markdown="1">
-		streamingContext.textFileStream(dataDirectory)
+		streamingContext.textFileStream(dataDirectory, depth)
     </div>
     </div>
 
-	Spark Streaming will monitor the directory `dataDirectory` and process any files created in that directory (files written in nested directories not supported). Note that
+	Spark Streaming will monitor the directory `dataDirectory`, the `depth` is default 1 and process any files created in that directory. If supported files written in nested directories, set the `depth` is greater than 1. Note that
```
Comment (srowen) [inline docs/streaming-programming-guide.md]: I think this needs to be rewritten, perhaps like:

```
Spark Streaming will monitor the directory `dataDirectory` and process any files created in that directory (files written in nested directories not supported). It can also monitor files in subdirectories by setting the optional `depth` parameter to a value greater than 1.
```

File: docs/streaming-programming-guide.md [hunk 4]
```diff
@@ -659,6 +659,7 @@ methods for creating DStreams from files and Akka actors as input sources.
      + Once moved, the files must not be changed. So if the files are being continuously appended, the new data will not be read.
 
 	For simple text files, there is an easier method `streamingContext.textFileStream(dataDirectory)`. And file streams do not require running a receiver, hence does not require allocating cores.
+	If Spark Streaming monitor the directory in nested directories, there is an easier method `streamingContext.textFileStream(dataDirectory, depth)`.
```
Comment (srowen) [inline docs/streaming-programming-guide.md]: I don't think this is necessary here.

File: examples/src/main/python/streaming/hdfs_wordcount.py [hunk 5]
```diff
@@ -39,7 +39,7 @@
     sc = SparkContext(appName="PythonStreamingHDFSWordCount")
     ssc = StreamingContext(sc, 1)
 
-    lines = ssc.textFileStream(sys.argv[1])
+    lines = ssc.textFileStream(sys.argv[1], 1)
```
Comment (srowen) [inline examples/src/main/python/streaming/hdfs_wordcount.py]: Same, I'd leave this as-is.

File: python/pyspark/streaming/context.py [hunk 6]
```diff
@@ -242,14 +242,14 @@ def socketTextStream(self, hostname, port, storageLevel=StorageLevel.MEMORY_AND_
         return DStream(self._jssc.socketTextStream(hostname, port, jlevel), self,
                        UTF8Deserializer())
 
-    def textFileStream(self, directory):
+    def textFileStream(self, directory, depth):
```
Comment (srowen) [inline python/pyspark/streaming/context.py]: We need to keep the existing method without the depth param right? or else the API changes.

File: streaming/src/main/scala/org/apache/spark/streaming/api/java/JavaStreamingContext.scala [hunk 7]
```diff
@@ -204,9 +204,10 @@ class JavaStreamingContext(val ssc: StreamingContext) extends Closeable {
    * monitored directory by "moving" them from another location within the same
    * file system. File names starting with . are ignored.
    * @param directory HDFS directory to monitor for new file
+   * @param depth Searching depth of directory
    */
-  def textFileStream(directory: String): JavaDStream[String] = {
-    ssc.textFileStream(directory)
+  def textFileStream(directory: String, depth: Int = 1): JavaDStream[String] = {
```
Comment (srowen) [inline streaming/src/main/scala/org/apache/spark/streaming/api/java/JavaStreamingContext.scala]: This won't work for Java API compatibility, I think. I think we need two methods, one of which has a non-optional new depth param.

File: streaming/src/main/scala/org/apache/spark/streaming/dstream/FileInputDStream.scala [hunk 8]
```diff
@@ -34,8 +34,10 @@ import org.apache.spark.util.{TimeStampedHashMap, Utils}
  * This class represents an input stream that monitors a Hadoop-compatible filesystem for new
  * files and creates a stream out of them. The way it works as follows.
  *
- * At each batch interval, the file system is queried for files in the given directory and
- * detected new files are selected for that batch. In this case "new" means files that
+ * At each batch interval, Use `depth` to find files in the directory recursively,
```
Comment (srowen) [inline streaming/src/main/scala/org/apache/spark/streaming/dstream/FileInputDStream.scala]: This also needs rewording, it doesn't quite make sense. Maybe use similar text to what I suggested above.

File: streaming/src/test/java/org/apache/spark/streaming/JavaAPISuite.java [hunk 9]
```diff
@@ -1739,7 +1739,11 @@ public Integer call(String s) throws Exception {
   // Java arguments and assign it to a JavaDStream without producing type errors. Testing of the
   // InputStream functionality is deferred to the existing Scala tests.
   @Test
-  public void testSocketTextStream() {
+  public void testSocketTextStream(
+
+
```
Comment (srowen) [inline streaming/src/test/java/org/apache/spark/streaming/JavaAPISuite.java]: Undo this change?

Comment (tdas) [inline streaming/src/test/java/org/apache/spark/streaming/JavaAPISuite.java]: Yes, this is a good functionality to revive. Thanks Sean

File: streaming/src/main/scala/org/apache/spark/streaming/dstream/FileInputDStream.scala [hunk 10]
```diff
@@ -173,8 +180,48 @@ class FileInputDStream[K, V, F <: NewInputFormat[K,V]](
...
+        }
+      }
+
+      val path = if (lastFoundDirs.isEmpty) Seq(fs.getFileStatus(directoryPath))
+      else {
+        lastFoundDirs.filter { path =>
+          // If the mod time of directory is more than ignore time, no new files in this directory.
+          try {
+            val status = fs.getFileStatus(path)
+            if (status != null && status.getModificationTime > modTimeIgnoreThreshold) true
```
Comment (srowen) [inline streaming/src/main/scala/org/apache/spark/streaming/dstream/FileInputDStream.scala]: You can get rid of the entire `if` since you're saying `if (x) true else false`. Just write `status != null && status.getModificationTime > modTimeIgnoreThreshold`

File: streaming/src/main/scala/org/apache/spark/streaming/dstream/FileInputDStream.scala [hunk 11]
```diff
@@ -173,8 +180,48 @@ class FileInputDStream[K, V, F <: NewInputFormat[K,V]](
...
+            else false
+          }
```
Comment (srowen) [inline streaming/src/main/scala/org/apache/spark/streaming/dstream/FileInputDStream.scala]: Tiny, but I think our style is to write `} catch {` and to put an `else` block on its own new line generally.

File: streaming/src/test/scala/org/apache/spark/streaming/InputStreamsSuite.scala [hunk 12]
```diff
@@ -278,11 +294,14 @@ class InputStreamsSuite extends TestSuiteBase with BeforeAndAfter {
     }
   }
 
-  def testFileStream(newFilesOnly: Boolean) {
+  def testFileStream(newFilesOnly: Boolean, depth :Int = 1) {
```
Comment (srowen) [inline streaming/src/test/scala/org/apache/spark/streaming/InputStreamsSuite.scala]: This should be `depth: Int = 1`. There is at least another instance above where `name : Type` is written; might as well standardize to `name: Type` everywhere.

File: streaming/src/main/scala/org/apache/spark/streaming/api/java/JavaStreamingContext.scala [hunk 13]
```diff
@@ -279,13 +289,14 @@ class JavaStreamingContext(val ssc: StreamingContext) extends Closeable {
    */
   def fileStream[K, V, F <: NewInputFormat[K, V]](
       directory: String,
+      depth: Int = 1,
```
Comment (srowen) [inline streaming/src/main/scala/org/apache/spark/streaming/api/java/JavaStreamingContext.scala]: This breaks the Java API since there is no longer a version without a depth param in Java right?

File: streaming/src/main/scala/org/apache/spark/streaming/api/java/JavaStreamingContext.scala [hunk 14]
```diff
@@ -309,12 +321,13 @@ class JavaStreamingContext(val ssc: StreamingContext) extends Closeable {
       vClass: Class[V],
       fClass: Class[F],
       filter: JFunction[Path, JBoolean],
-      newFilesOnly: Boolean): JavaPairInputDStream[K, V] = {
+      newFilesOnly: Boolean,
+      depth: Int = 1): JavaPairInputDStream[K, V] = {
```
Comment (srowen) [inline streaming/src/main/scala/org/apache/spark/streaming/api/java/JavaStreamingContext.scala]: Same issue

File: project/MimaExcludes.scala [hunk 15]
```diff
@@ -83,7 +83,16 @@ object MimaExcludes {
             // SPARK-2757
             ProblemFilters.exclude[IncompatibleResultTypeProblem](
               "org.apache.spark.streaming.flume.sink.SparkAvroCallbackHandler." +
-                "removeAndGetProcessor")
+                "removeAndGetProcessor"),
+            // SPARK-3586
```
Comment (srowen) [inline project/MimaExcludes.scala]: I think we should not have any MiMa failures here. The Java errors are correct and need to be avoided. I'd have to look again at why the Scala API is changed by this.

File: python/pyspark/streaming/context.py [hunk 16]
```diff
@@ -260,6 +260,12 @@ def textFileStream(self, directory):
         """
         return DStream(self._jssc.textFileStream(directory), self, UTF8Deserializer())
```
Comment (srowen) [inline python/pyspark/streaming/context.py]: Should this just call the other method with arg 1 instead of duplicating the logic?

File: python/pyspark/streaming/tests.py [hunk 17]
```diff
@@ -446,7 +446,7 @@ def test_queue_stream(self):
     def test_text_file_stream(self):
         d = tempfile.mkdtemp()
         self.ssc = StreamingContext(self.sc, self.duration)
-        dstream2 = self.ssc.textFileStream(d).map(int)
+        dstream2 = self.ssc.textFileStream(d, 1).map(int)
```
Comment (srowen) [inline python/pyspark/streaming/tests.py]: I'd omit ", 1" since that is the 'default'. This test actually does not change.

**Comment by wangxiaojing** [comment]
@srowen This PR  does not affect ·org.apache.spark.JavaAPISuite.aggregateByKey·. Please test this again, thanks.

**Comment by wangxiaojing** [comment]
@srowen Can we test this again please ?

**Comment by srowen** [comment]
This is getting better. I am still concerned about the MiMa excludes. There should be no API changes, so, there should be no excludes. Those are still in this PR.

The additions to the Scala API are not using optional arguments anymore. I think it's most natural to create an option depth argument on the most "complete" version of each API method. It is not necessary in Scala to overload a method to add an optional argument, and it is not necessary to add this argument to every version of the method.

That is, I think I would expect to see one, maybe two versions of each Scala API method add a `depth: Int = 1` parameter, only.

In Java, there are no optional params (even though we're writing in Scala) but I also believe it is not desirable to overload every single method. The one with the most args can take the overload.

File: streaming/src/main/scala/org/apache/spark/streaming/api/java/JavaStreamingContext.scala [hunk 18]
```diff
@@ -293,6 +302,34 @@ class JavaStreamingContext(val ssc: StreamingContext) extends Closeable {
...
+   * @param kClass class of key for reading HDFS file
+   * @param vClass class of value for reading HDFS file
+   * @param fClass class of input format for reading HDFS file
+   * @tparam K Key type for reading HDFS file
+   * @tparam V Value type for reading HDFS file
+   * @tparam F Input format for reading HDFS file
+   */
+  def fileStream[K, V, F <: NewInputFormat[K, V]](
+      directory: String,
+      depth: Int,
```
Comment (srowen) [inline streaming/src/main/scala/org/apache/spark/streaming/api/java/JavaStreamingContext.scala]: This looks out of order, but, I question the need for this overload anyway. Maybe @tdas has an opinion. I'd like to keep the scope of this change much more limited to adding a depth param to 1 version of each method (maybe 2 if it really made sense) in each language.

**Comment by wangxiaojing** [comment]
@srowen There is two possible solution,no API change.
     1.Adding a switch Whether to  monitor files in subdirectories,Defaults to false will monitor the directory `dataDirectory` , If set to true,monitor files in subdirectories. 
   2.Adding a  configuration `streaming.monitor.directory.depth` to  control the searching depth of directories,Defaults to 1 will monitor the directory `dataDirectory` , If set to  greater than 1, monitor files in subdirectories.
    Can you give some advice?

**Comment by srowen** [comment]
I don't think an additional switch helps; a depth parameter is the right way to specify the depth. I also don't think it's a configuration parameter as it is specific to the method call.

What I'm suggesting is to simply not have so many overloads. A method like `fileStream` need only add the depth parameter to one of its overloads, not all. It can go on the overload with the most arguments.

**Comment by srowen** [comment]
Do you mind closing this PR? It seems like it is not being worked on.

**Comment by lisendong** [comment]
@wangxiaojing would you continue work on this PR, I need this feature:-)

**Comment by erfangc** [comment]
This feature would definitely be helpful. Thanks to @wangxiaojing and whoever continuing to work on PR!

**Comment by zsxwing** [comment]
@wangxiaojing could you update this PR? It conflicts with master

File: streaming/src/main/scala/org/apache/spark/streaming/dstream/FileInputDStream.scala [hunk 19]
```diff
@@ -183,8 +189,46 @@ class FileInputDStream[K, V, F <: NewInputFormat[K,V]](
...
+            // If the directory don't find, remove the directory from `lastFoundDirs`
+            case e: FileNotFoundException =>
+              lastFoundDirs.remove(path)
+              false
+          }
+        }
+      }.flatMap(fs.listStatus(_)).toSeq
+
+      val newFiles = path.flatMap(dfs(_)).map(_.getPath.toString).toArray
+      val timeTaken = System.currentTimeMillis - lastNewFileFindingTime
```
Comment (zsxwing) [inline streaming/src/main/scala/org/apache/spark/streaming/dstream/FileInputDStream.scala]: Could you change `System.currentTimeMillis` to `clock.getTimeMillis()`?

File: streaming/src/main/scala/org/apache/spark/streaming/dstream/FileInputDStream.scala [hunk 20]
```diff
@@ -183,8 +189,46 @@ class FileInputDStream[K, V, F <: NewInputFormat[K,V]](
...
       }
-      val newFiles = fs.listStatus(directoryPath, filter).map(_.getPath.toString)
-      val timeTaken = clock.getTimeMillis() - lastNewFileFindingTime
+      val directoryDepth = fs.getFileStatus(directoryPath).getPath.depth()
+
+      // Nested directories to find new files.
+      def dfs(status: FileStatus): List[FileStatus] = {
+        val path = status.getPath
+        val depthFilter = depth + directoryDepth - path.depth()
+        if (status.isDir) {
```
Comment (zsxwing) [inline streaming/src/main/scala/org/apache/spark/streaming/dstream/FileInputDStream.scala]: Could you move `val depthFilter = depth + directoryDepth - path.depth()` into the `if` block? It can avoid calling `path.depth()` for all files.

File: streaming/src/main/scala/org/apache/spark/streaming/dstream/FileInputDStream.scala [hunk 21]
```diff
@@ -223,6 +266,11 @@ class FileInputDStream[K, V, F <: NewInputFormat[K,V]](
    */
   private def isNewFile(path: Path, currentTime: Long, modTimeIgnoreThreshold: Long): Boolean = {
     val pathStr = path.toString
+    // Reject file if it start with _
+    if (path.getName().startsWith("_")) {
```
Comment (zsxwing) [inline streaming/src/main/scala/org/apache/spark/streaming/dstream/FileInputDStream.scala]: Could you add this filter to `FileInputDStream.defaultFilter`? 

@tdas is it a bug that `FileInputDStream.defaultFilter` only ignores files that start with `.`? Hadoop uses the following path filter (FileInputFormat.hiddenFileFilter):
```Java
    private static final PathFilter hiddenFileFilter = new PathFilter() {
        public boolean accept(Path p) {
            String name = p.getName();
            return !name.startsWith("_") && !name.startsWith(".");
        }
    };
```
Since SparkContext.textFile uses `TextInputFormat`, it uses this filter and ignores files that start with "-" or ".". Should Streaming be consistent with Core?

Comment (tdas) [inline streaming/src/main/scala/org/apache/spark/streaming/dstream/FileInputDStream.scala]: Its good to be consistent with the core, however, I am worried that if we
add another filter, it will silently break existing workloads that may be
using files that start with "_"

**Comment by andrewor14** [comment]
Hi @wangxiaojing it seems that #6588 is an updated version of this PR. Would you mind closing this patch since it no longer merges cleanly with master?

**Comment by wangxiaojing** [comment]
@andrewor14  ok.
